dialogue-core/src/main/java/com/palantir/dialogue/core/Balanced.java

```diff
@@ -0,0 +1,173 @@
+ * being served and also how many failures have been seen in the last few seconds.
+ *
+ * This is intended to be a strict improvement over Round Robin and Random Selection which can leave fast servers
+ * underutilized, as it sends the same number to both a slow and fast node. It is *not* appropriate for transactional
+ * workloads (where n requests must all land on the same server) or scenarios where cache warming is very important.
+ * {@link PinUntilErrorChannel} remains the best choice for these.
+ */
+final class Balanced implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(Balanced.class);
+
+    /**
+     * We'd like to remember failures for a long time, but this would increase CPU on a hot codepath as we compute
+     * the number of failures for each channel in order to sort. See {@link ChannelWithStats#score}.
+     */
+    private static final Duration FAILURE_MEMORY = Duration.ofSeconds(5);
+
+    /**
+     * This comparator is a little risky because the data can change while we're sorting. In practise the sort
+     * finishes a lot quicker than network requests do, and we don't mind if it's not exactly right.
+     */
+    @VisibleForTesting
+    static final Comparator<ChannelWithStats> BY_SCORE = Comparator.comparing(ChannelWithStats::score);
+
+    private final ImmutableList<ChannelWithStats> channels;
+    private final Random random;
+    private final CodahaleClock clock;
+
+    Balanced(ImmutableList<LimitedChannel> channels, Random random, Ticker ticker) {
+        Preconditions.checkState(channels.size() >= 2, "At least two channels required");
+        this.random = random;
+        this.clock = new CodahaleClock(ticker);
+        this.channels = channels.stream()
+                .map(channel -> new ChannelWithStats(channel, clock))
+                .collect(ImmutableList.toImmutableList());
+
+        log.debug("Initialized", SafeArg.of("count", channels.size()), UnsafeArg.of("channels", channels));
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        if (log.isDebugEnabled()) {
+            log.debug(
+                    "Channel scores {}",
+                    SafeArg.of("scores", Lists.transform(channels, ChannelWithStats::score)),
+                    SafeArg.of("inflight", Lists.transform(channels, ChannelWithStats::inflight)));
+        }
+
+        // pre-shuffling is pretty important here, otherwise when there are no requests in flight, we'd
+        // *always* prefer the first channel of the list, leading to a higher overall load.
+        List<ChannelWithStats> preShuffled = shuffleImmutableList(channels, random);
+
+        // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
+        // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
+        return preShuffled.stream()
+                .sorted(BY_SCORE)
```

Comment:
I think this sort can throw because score will be modified by concurrent requests, list[a] > list[b] when first validated, but not the second time.

Reply:
Hmmm, I guess we could copy the mutable things into an immutable list and then sort that?

Reply:
Yep, that should do it

Reply:
Have now implemented an immutable snapshot.